backend/src/ast/converters/scratch/helpers.ts

```diff
@@ -0,0 +1,322 @@
+  inputs?: {
+    [key in TKey]?: BlockInputValue | string;
+  };
+  __children: NonHatBlockTree[];
+}): ExpressionNode => ({
+  nodeType: AstNodeType.code,
+  codeType: CodeNodeType.expression,
+  expressionType: ExpressionNodeType.functionCall,
+  name: block.opcode,
+  arguments: createArgumentsFromInputsAndFields(block),
+});
+
+export const createVariableExpressionBlock = (
+  variableName: string,
+): ExpressionNode => ({
+  nodeType: AstNodeType.code,
+  codeType: CodeNodeType.expression,
+  expressionType: ExpressionNodeType.variable,
+  name: variableName,
+});
+
+export const createOperatorExpressionBlock = <TKey extends string>(block: {
+  opcode: string;
+  fields?: Block["fields"];
+  inputs?: {
+    [key in TKey]?: BlockInputValue | string;
+  };
+  __children: NonHatBlockTree[];
+}): ExpressionNode => ({
+  nodeType: AstNodeType.code,
+  codeType: CodeNodeType.expression,
+  expressionType: ExpressionNodeType.operator,
+  operator: block.opcode,
+  operands: createArgumentsFromInputsAndFields(block),
+});
+
+export const getEventParameters = (
+  block: BlockTreeWithEventHatRoot,
+): ExpressionNode[] => {
+  return match(block)
+    .returnType<ExpressionNode[]>()
+    .with(P.when(isWhenBackdropSwitchesToBlock), (block) => [
+      createLiteralNode("string", block.fields.BACKDROP[0]),
+    ])
+    .with(P.when(isWhenBroadcastReceivedBlock), (block) => [
+      createLiteralNode("string", block.fields.BROADCAST_OPTION[0]),
+    ])
+    .with(
+      P.when(isWhenGreaterThanBlock),
+      (block: WhenGreaterThanBlock & BlockTreeWithEventHatRoot) => [
+        convertInputsToExpression(block, "VALUE"),
+        createLiteralNode("string", block.fields.WHENGREATERTHANMENU[0]),
+      ],
+    )
+    .with(P.when(isWhenKeyPressedBlock), (block) => [
+      createLiteralNode("string", block.fields.KEY_OPTION[0]),
+    ])
+    .with(
+      P.when(isWhenFlagClickedBlock),
+      P.when(isWhenStageIsClickedBlock),
+      P.when(iswhenThisSpriteIsClickedBlock),
+      P.when(isStartAsCloneBlock),
+      () => [],
+    )
+    .exhaustive();
+};
+
+export const getParametersFromFields = (
+  fields: Block["fields"],
+): ({ parameterName: string } & (
+  | { value: string }
+  | { variableName: string }
+))[] => {
+  if (!fields) {
+    return [];
+  }
+
+  return Object.entries(fields).map(([parameterName, value]) => {
+    if (!Array.isArray(value)) {
+      throw new Error(
+        `Unexpected field value for name '${parameterName}': '${JSON.stringify(value)}'`,
+      );
+    }
+
+    // while I could not find any documentation on this, it seems that fields are always an array
+    // consisting of two elements where the first one is the value or variable name
+    // and the second one is either null of the value is a literal or a block id if the value is a variable
+    // however for variables the variable name is enough for us and we don't need the block id
+    // therefore we only take the first element
```

Comment:
Could we turn this into actual code? I.e. a pattern matching on `(value, null)` or `(variableName, blockId)`, and everything else throws ? We should still keep the warning that this is based on our understanding of undocumented Scratch behaviours, but it would be much more readable.